File: 2019/03/ruby_day03/03.rb
#!/usr/bin/env ruby
$DIRECTIONS = {
  'R' => Complex(1, 0),
  'L' => Complex(-1, 0),
  'U' => Complex(0, 1),
  'D' => Complex(0, -1)
}

# Input (String): "L1003,U125,L229,U421"
# Output (Wire): [["L", 1003], ["U", 125], ["L", 229], ["U", 421]]
def parse_wire(line)
  line.split(',').map do |pair|
    compass, l = pair.match(/^(\w)(\d+)$/).captures
    [compass, l.to_i]
  end
end

# Input (filename: String) -> Output (List of Wires)
def parse(filename)
  File.readlines(filename).map { |line| parse_wire(line) }
end

def manhattan(coord)
  coord.real.abs + coord.imaginary.abs
end

def total_steps(grid_value)
  grid_value.map { |info| info[:total_steps] }.sum
end

def part12(data)
  grid = Hash.new(nil)
  data.each_with_index do |wire, wire_index|
    location = Complex(0, 0)
    total_steps = 0
    wire.each do |step|
      delta = $DIRECTIONS[step[0]]
      1.upto(step[1]) do |i|
        location += delta
        total_steps += 1
        info = { wire_index: wire_index, total_steps: total_steps }
        # puts "i #{i} location #{location} wire #{wire_index}"

        if grid[location] == nil
          grid[location] = [info]
        elsif !grid[location].any? { |thing| thing[:wire_index] == wire_index }
          grid[location].push(info)
        end
      end
    end
  end

  intersections = grid.keys.filter { |k| grid[k].length >= 2 }
  answer1 = intersections.map { |k| manhattan(k) }.min
  answer2 = intersections.map { |k| total_steps(grid[k]) }.min

  [answer1, answer2]
end

data = parse('../input.txt')
answers = part12(data)
puts 'Part 1:'
puts answers[0]
puts 'Part 2:'
puts answers[1]

File: 2019/03/python_day03/03.py
#!/usr/bin/env python
from collections import defaultdict

DIRECTIONS = {
    "R": complex(1, 0),
    "L": complex(-1, 0),
    "U": complex(0, 1),
    "D": complex(0, -1),
}


# filename -> [ wires ]
# wire example: [ ("U", 5), ("D", 40), ... ]
def parse(filename):
    return [parse_line(line) for line in open(filename).readlines()]


# string -> wire
# wire example: [ ("U", 5), ("D", 40), ... ]
def parse_line(line):
    return [parse_step(step) for step in line.strip().split(",")]


# string -> step
# step example: ("R", 10)
def parse_step(step):
    direction = step[0]
    distance = int(step[1:])
    return (direction, distance)


def manhattan(coord):
    return abs(coord.real) + abs(coord.imag)


def solve(wires):
    grid_wire_ids = defaultdict(list)
    grid_steps = defaultdict(list)
    wire_id = 0
    for wire in wires:
        wire_id += 1
        steps = 0
        location = complex(0, 0)
        for (direction, distance) in wire:
            for _ in range(distance):
                location += DIRECTIONS[direction]
                steps += 1
                if wire_id not in grid_wire_ids[location]:
                    grid_wire_ids[location].append(wire_id)
                    grid_steps[location].append(steps)

    intersections = [coord for coord in grid_wire_ids if len(grid_wire_ids[coord]) > 1]
    answer1 = int(min(manhattan(coord) for coord in intersections))
    answer2 = min(sum(grid_steps[coord]) for coord in intersections)
    return (answer1, answer2)


data = parse("../input.txt")
print(solve(data))
